test/integration/apiserver/admissionwebhook/reinvocation_test.go

```diff
@@ -259,48 +314,28 @@ func TestWebhookReinvocationPolicy(t *testing.T) {
 	}
 }
 
-func registerWebhook(t *testing.T, client clientset.Interface, name, endpoint string, reinvocationPolicy *registrationv1beta1.ReinvocationPolicyType, objectSelector *metav1.LabelSelector) func() {
-	fail := admissionv1beta1.Fail
-	hook, err := client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Create(&admissionv1beta1.MutatingWebhookConfiguration{
-		ObjectMeta: metav1.ObjectMeta{Name: name},
-		Webhooks: []admissionv1beta1.MutatingWebhook{{
-			Name: name,
-			ClientConfig: admissionv1beta1.WebhookClientConfig{
-				URL:      &endpoint,
-				CABundle: localhostCert,
-			},
-			Rules: []admissionv1beta1.RuleWithOperations{{
-				Operations: []admissionv1beta1.OperationType{admissionv1beta1.OperationAll},
-				Rule:       admissionv1beta1.Rule{APIGroups: []string{"*"}, APIVersions: []string{"*"}, Resources: []string{"*/*"}},
-			}},
-			ObjectSelector:          objectSelector,
-			FailurePolicy:           &fail,
-			ReinvocationPolicy:      reinvocationPolicy,
-			AdmissionReviewVersions: []string{"v1beta1"},
-		}},
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	tearDown := func() {
-		err := client.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Delete(hook.GetName(), &metav1.DeleteOptions{})
-		if err != nil {
-			t.Fatal(err)
-		}
-	}
-	return tearDown
-}
-
 type invocationRecorder struct {
 	mu     sync.Mutex
+	upCh   chan struct{}
+	upOnce sync.Once
 	counts map[string]int
 }
 
-func (i *invocationRecorder) Reset() {
+// Reset zeros out all counts and returns a channel that is closed when the first admission of the
+// marker object is received.
+func (i *invocationRecorder) Reset() chan struct{} {
 	i.mu.Lock()
 	defer i.mu.Unlock()
 	i.counts = map[string]int{}
+	i.upCh = make(chan struct{})
+	i.upOnce = sync.Once{}
+	return i.upCh
+}
+
+func (i *invocationRecorder) MarkerReceived() {
+	i.upOnce.Do(func() {
```

Comment:
should happen under `i.mu.Lock()`, right? so the `i.upOnce` assignment on line 331 doesn't race with the access here?

Reply:
oops, yes, fixing